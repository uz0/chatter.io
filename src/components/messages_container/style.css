.messages {
  display: flex;
  flex-direction: column;
}

.messages .header {
  flex-shrink: 0;
  box-shadow: 0 -1px var(--divider-color) inset;
}

.messages .list {
  flex-grow: 1;
  overflow-x: hidden;
  overflow-y: auto;
  min-height: 0;
  display: flex;
  flex-direction: column-reverse;
  align-items: flex-start;
  padding-left: var(--layout-padding);
  padding-right: var(--layout-padding);
  padding-bottom: 20px;
}

.messages .list .item {
  flex-shrink: 0;
}

.messages .list .item:last-child {
  margin-top: auto;
  padding-top: 20px;
}

.messages .list .message + .message {
  margin-bottom: 2px;
}

.messages .input {
  flex-shrink: 0;
}
