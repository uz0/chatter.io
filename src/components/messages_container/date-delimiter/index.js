import React, { Component } from 'react';
import classnames from 'classnames/bind';
import style from './style.css';

const cx = classnames.bind(style);

class DateDelimiter extends Component {
  render() {
    return <p className={cx('date', this.props.className)}>Today</p>;
  }
}

export default DateDelimiter;
