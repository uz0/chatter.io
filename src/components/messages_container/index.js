import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import compose from 'recompose/compose';
import get from 'lodash/get';
import map from 'lodash/map';
import groupBy from 'lodash/groupBy';
import moment from 'moment';
import classnames from 'classnames/bind';
import { withNamespaces } from 'react-i18next';
import Header from './header';
import MessageInput from './input';
import DateDelimiter from './date-delimiter';
import XtagDelimiter from './xtag-delimiter';
import UnreadDelimiter from './unread-delimiter';
import MessageItem from '@/components/message-item';
import Loading from '@/components/loading';
import { withDetails } from '@/hoc';
import { api } from '@';
import { uid, getChatName } from '@/helpers';
import { actions as messagesActions } from '@/store/messages';
import style from './style.css';

const cx = classnames.bind(style);

class Messages extends Component {
  state = {
    isMessagesLoading: false,
  };

  getGroupedMessages = () => {
    if (!this.props.details) {
      return null;
    }

    if (!this.props.chatIds || !this.props.chatIds.isLoaded) {
      return null;
    }

    const groupedByDate = groupBy(
      map(this.props.chatIds.list, id => this.props.messages_list[id]),
      message => moment(message.created_at).format('YYYY-MM-DD'),
    );

    let array = [];
    const dates = Object.keys(groupedByDate).sort();

    dates.forEach(key => {
      array.push({ type: 'dateDelimiter', date: key });

      let messages = [];

      groupedByDate[key].reverse().forEach(message => {
        if (message.xtag) {
          if (messages.length > 0) {
            array.push({ type: 'messages', messages_ids: messages });
          }

          array.push({ type: 'xtagDelimiter', message_id: message.id });
          messages = [];
          return;
        }

        messages.push(message.id);
      });

      if (messages.length > 0) {
        array.push({ type: 'messages', messages_ids: messages });
      }
    });

    return array;
  };

  loadMessages = props => {
    this.setState({ isMessagesLoading: true });

    api.getMessages({ subscription_id: props.details.id, limit: 100 }).then(data => {
      this.setState({ isMessagesLoading: false });
      this.props.loadMessages({chatId: props.details.id, list: data.messages, isLoaded: true});
    });
  }

  componentDidMount() {
    const isMessagesLoaded = get(this.props, 'chatIds.isLoaded', false);

    if (this.props.details && !isMessagesLoaded) {
      this.loadMessages(this.props);
    }
  }

  componentWillReceiveProps(nextProps) {
    const isMessagesLoaded = get(nextProps, 'chatIds.isLoaded', false);
    const isChatChanged = this.props.details && nextProps.details && this.props.details.id !== nextProps.details.id;

    // если страница загружается сразу с открытым чатом, details еще не успевает прийти, ловим тут
    if (!this.props.details && nextProps.details && !isMessagesLoaded) {
      this.loadMessages(nextProps);
    }

    // если перешли в другой чат из существующего и в нем не прогружены сообщения
    if (isChatChanged && !isMessagesLoaded) {
      this.loadMessages(nextProps);
    }
  }

  // shouldComponentUpdate(nextProps) {
  //   console.log(this.props);
  //   console.log(nextProps);
  //   return true;
  // }

  render() {
    // console.log(123);
    const groupedMessages = this.getGroupedMessages() || [];

    return <div className={cx('messages', this.props.className)}>
      {this.props.details &&
        <Header
          title={getChatName(this.props.details)}
          count={this.props.details.group.participants.length}
          className={style.header}
        />
      }

      <div className={style.list}>
        {groupedMessages &&
          groupedMessages.reverse().map(grouped => <Fragment key={uid()}>
            {grouped.type === 'unreadDelimiter' &&
              <UnreadDelimiter className={cx('item')} />
            }

            {grouped.type === 'xtagDelimiter' &&
              <XtagDelimiter id={grouped.message_id} className={cx('item')} />
            }

            {grouped.type === 'dateDelimiter' &&
              <DateDelimiter date={grouped.date} className={cx('item')} />
            }

            {grouped.type === 'messages' &&
              grouped.messages_ids.reverse().map(message_id => <MessageItem
                key={message_id}
                id={message_id}
                className={cx('message', 'item')}
              />)
            }
          </Fragment>)
        }

        {groupedMessages.length === 0 &&
          <p className={style.empty}>{this.props.t('there_is_no_messages')}</p>
        }

        <Loading isShown={this.state.isMessagesLoading} className={style.loading} />
      </div>

      {this.props.details &&
        <MessageInput subscription_id={this.props.details.id} className={style.input} />
      }
    </div>;
  }
}

export default compose(
  withDetails,
  withNamespaces('translation'),

  connect(
    state => ({
      messages_list: state.messages.list,
    }),

    {
      loadMessages: messagesActions.loadMessages,
    },
  ),

  connect(
    (state, props) => ({
      chatIds: props.details ? state.messages.chatIds[props.details.id] : null,
    }),
  ),
)(Messages);
