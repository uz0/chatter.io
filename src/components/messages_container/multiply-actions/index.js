import React, { Component } from 'react';
import compose from 'recompose/compose';
import { connect } from 'react-redux';
import Button from '@/components/button';
import { actions as messagesActions } from '@/store/messages';
import classnames from 'classnames/bind';
import style from './style.css';

import actions from './actions';

const cx = classnames.bind(style);

class MultiplyActions extends Component {
  componentWillReceiveProps(nextProps) {
    const isChatChanged = this.props.subscription_id !== nextProps.subscription_id;

    if (isChatChanged) {
      this.props.resetCheckedMessages();
    }
  }

  componentWillUnmount() {
    this.props.resetCheckedMessages();
  }

  render() {
    if (!this.props.isCheckShown) {
      return null;
    }

    return <div className={cx('multiply', this.props.className)}>
      <Button appearance="_basic-primary" text="Forward" className={style.button} />
      <Button appearance="_basic-primary" text="Delete" className={style.button} onClick={this.props.deleteMessages} />
      <Button appearance="_basic-divider" text="Cancel" className={style.button} onClick={this.props.resetCheckedMessages} />
    </div>;
  }
}

export default compose(
  connect(
    state => ({
      isCheckShown: state.messages.checked_ids.length > 0,
    }),

    {
      resetCheckedMessages: messagesActions.resetCheckedMessages,
      deleteMessages: actions.deleteMessages,
    },
  ),
)(MultiplyActions);
