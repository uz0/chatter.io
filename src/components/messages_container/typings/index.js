import React, { Component } from 'react';
import { connect } from 'react-redux';
import compose from 'recompose/compose';
import isEmpty from 'lodash/isEmpty';
import { withTypings } from '@/hoc';
import style from './style.css';

class Typings extends Component {
  componentWillReceiveProps(nextProps) {
    const isTypingsHasBeenShown = isEmpty(this.props.typings) && !isEmpty(nextProps.typings);

    if (isTypingsHasBeenShown) {
      const messagesScrollElement = document.getElementById('messages-scroll');
      const isScrolledToBottom = messagesScrollElement && messagesScrollElement.scrollTop === (messagesScrollElement.scrollHeight - messagesScrollElement.offsetHeight);

      if (isScrolledToBottom) {
        setTimeout(() => messagesScrollElement.scrollTo(0, messagesScrollElement.scrollHeight));
      }
    }
  }

  render() {
    return !isEmpty(this.props.typings) &&
      <p className={style.typings}>{this.props.typings}</p>;
  }
}

export default compose(
  connect(
    (state, props) => ({
      details: state.subscriptions.list[props.subscription_id],
    }),
  ),

  withTypings(props => ({
    chat: props.details,
  })),
)(Typings);
