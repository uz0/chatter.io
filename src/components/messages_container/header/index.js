import React, { Component } from 'react';
import { connect } from 'react-redux';
import compose from 'recompose/compose';
import classnames from 'classnames/bind';
import SubscriptionAvatar from '@/components/subscription-avatar';
import Icon from '@/components/icon';
import { actions as modalActions} from '@/components/modal_container';
import { withRouter } from '@/hoc';
import { getChatName, getOpponentUser, getLastActive } from '@/helpers';
import style from './style.css';

const cx = classnames.bind(style);

class Header extends Component {
  showPanelContainer = () => this.props.toggleModal({ id: 'panel-container' });

  closeChat = () => {
    this.props.closeModal('panel-container');
    this.props.pushUrl('/chat');
  };

  getLastActive = chat => {
    if (chat.group.type !== 'private_chat') {
      return null;
    }

    const opponent = getOpponentUser(chat);

    if (!opponent) {
      return null;
    }

    const user = this.props.users_list[opponent.id];
    return getLastActive(user, () => this.forceUpdate());
  };

  render() {
    const name = getChatName(this.props.details);
    const count = this.props.details.group.participants.length;
    const isChatPrivate = this.props.details.group.type === 'private_chat';
    const lastActive = this.getLastActive(this.props.details);

    return <div className={cx('header', this.props.className)}>
      <button className={style.back} onClick={this.closeChat}>
        <Icon name="arrow-left" />
      </button>

      <SubscriptionAvatar subscription={this.props.details} className={style.avatar} />

      <div className={style.section}>
        <button onClick={this.showPanelContainer}>{name}</button>

        {!isChatPrivate &&
          <p className={style.count}>{count} people</p>
        }

        {isChatPrivate &&
          <p className={style.count}>{lastActive}</p>
        }
      </div>
    </div>;
  }
}

export default compose(
  withRouter,

  connect(
    (state, props) => ({
      details: state.subscriptions.list[props.chatId],
      users_list: state.users.list,
    }),

    {
      toggleModal: modalActions.toggleModal,
      closeModal: modalActions.closeModal,
    },
  ),
)(Header);
