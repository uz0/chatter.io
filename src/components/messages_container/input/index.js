import React, { Component } from 'react';
import compose from 'recompose/compose';
import { connect } from 'react-redux';
import { withNamespaces } from 'react-i18next';
import find from 'lodash/find';
import get from 'lodash/get';
import isEqual from 'lodash/isEqual';
import classnames from 'classnames/bind';
import Button from '@/components/button';
import Icon from '@/components/icon';
import Loading from '@/components/loading';
import Validators from '@/components/form/validators';
import throttle from 'lodash/throttle';
import CRC32 from 'crc-32';
import inputActions from './actions';
import Message from './message';
import { api } from '@';
import { actions as notificationActions } from '@/components/notification';
import { actions as messagesActions } from '@/store/messages';
import style from './style.css';

export { default as actions } from './actions';

const cx = classnames.bind(style);
const bytesSize = 64000;

class MessageInput extends Component {
  state = {
    attachment: null,
    upload_id: null,
    value: this.props.draft || '',
    isFileLoading: false,
  };

  onTextareaKeyDown = event => {
    if (event.keyCode === 13) {
      event.preventDefault();
    }
  };

  setNewLineTextarea = () => {
    const selection = this.textareaRef.selectionStart;
    const newValue = this.state.value.substring(0, selection) + '\n' + this.state.value.substring(selection, this.state.value.length);
    this.setState({ value: newValue });
    this.calcTextareaHeight();
    setTimeout(() => this.textareaRef.setSelectionRange(selection + 1, selection + 1));
  };

  mobileDocumentKeyDown = event => {
    if (event.keyCode === 13 && document.activeElement === this.textareaRef) {
      this.setNewLineTextarea();
    }
  };

  desktopDocumentKeyDown = event => {
    if (event.keyCode === 13 && event.shiftKey && document.activeElement === this.textareaRef) {
      this.setNewLineTextarea();
    }

    if (event.keyCode === 13 && !event.shiftKey && document.activeElement === this.textareaRef) {
      this.onSendButtonClick();
    }
  };

  handleDocumentKeyDown = event => {
    if (this.props.isMobile) {
      this.mobileDocumentKeyDown(event);
    } else {
      this.desktopDocumentKeyDown(event);
    }
  };

  attachFile = () => this.attachInputRef.click();

  resetAttachment = () => {
    this.attachInputRef.value = '';
    this.setState({ attachment: null, upload_id: null, isFileLoading: false });
  };

  createLinkToJsonArguments = object => {
    const a = document.createElement('a');
    const data = 'text/json;charset=utf-8,' + encodeURIComponent(JSON.stringify(object));
    a.setAttribute('href', `data:${data}`);
    a.setAttribute('download', 'data.json');
    document.body.appendChild(a);
  };

  getBlobBase = blob => new Promise(resolve => {
    const reader = new FileReader();

    reader.onloadend = event => {
      let binaryString = '';
      const bytes = new Uint8Array(event.target.result);

      for (let i = 0; i < bytes.byteLength; i++) {
        binaryString += String.fromCharCode(bytes[i]);
      }

      resolve(window.btoa(binaryString));
    };

    reader.readAsArrayBuffer(blob);
  });

  getFileChecksum = file => new Promise(resolve => {
    const reader = new FileReader();

    reader.onloadend = event => {
      const bytes = new Uint8Array(event.target.result);
      resolve(CRC32.buf(bytes) >>> 0);
    };

    reader.readAsArrayBuffer(file);
  });

  loadFileByChunks = async file => {
    try {
      this.setState({ isFileLoading: true });
      const firstResponse = await this.loadFirstChunk(file);
      await this.loadLastChunks(file, firstResponse);
      this.setState({ upload_id: firstResponse.upload_id, isFileLoading: false });
    } catch (error) {
      console.error(error);
      this.props.showNotification(error.text);
      this.resetAttachment();
    }
  };

  loadFirstChunk = async file => {
    let blob = file.slice(0, bytesSize);
    let chunk = await this.getBlobBase(blob);
    const checksum = await this.getFileChecksum(file);

    return api.attachmentByChunks({
      file_chunk: chunk,
      upload_id: null,
      file_size: file.size,
      file_checksum: checksum,
      file_name: file.name,
    });
  };

  loadLastChunks = async (file, firstResponse) => {
    const checksum = await this.getFileChecksum(file);

    for (let i = bytesSize; i <= file.size; i += bytesSize) {
      let blob = file.slice(i, i + bytesSize);
      let chunk = await this.getBlobBase(blob);

      await api.attachmentByChunks({
        file_chunk: chunk,
        upload_id: firstResponse.upload_id,
        file_size: file.size,
        file_checksum: checksum,
        file_name: file.name,
      });
    }
  };

  onAttachFileChange = event => {
    const file = event.target.files[0];

    if (!file) {
      return;
    }

    if (file.size === 0) {
      this.props.showNotification(this.props.t('you_cannot_upload_empty_files'));
      return;
    }

    const reader = new FileReader();

    reader.onloadend = () => {
      if (Validators.fileMaxSize(200000)(file)) {
        this.loadFileByChunks(file);
      }

      this.setState({
        attachment: {
          byte_size: file.size,
          content_type: file.type,
          preview: reader.result,
          url: reader.result,
        },
      });
    };

    reader.readAsDataURL(file);
  };

  getFilteredMessage = value => {
    if (!value) {
      return '';
    }

    let text = value.replace(/\r|\n|\r\n/g, '<br />');

    if (text[0] === ' ') {
      text = text.substring(1);
    }

    if (text[text.length - 1] === ' ') {
      text = text.substring(0, text.length - 1);
    }

    return text;
  };

  parseMentions = text => {
    if (!text) {
      return null;
    }

    let mentions = [];

    this.props.users_ids.forEach(id => {
      const user = this.props.users_list[id];

      if (user.nick && text.match(`@${user.nick}`) && !find(mentions, {user_id: user.id})) {
        mentions.push({ user_id: user.id, text: user.nick });
      }
    });

    return mentions || null;
  };

  editMessage = ({ text, attachment, upload_id }) => {
    this.props.updateMessage({
      ...text ? {text} : {},
      ...attachment ? {attachment} : {},
      ...upload_id ? {upload_id} : {},
      ...this.props.edit_message_id ? {edit_message_id: this.props.edit_message_id} : {},
    });

    this.props.clearEditMessage();
    this.setState({ value: '', attachment: null, upload_id: null });
    setTimeout(() => this.calcTextareaHeight());
  };

  sendMessage = ({ text, attachment, mentions, upload_id }) => {
    this.props.sendMessage({
      ...text ? {text} : {},
      ...attachment ? {attachment} : {},
      ...mentions ? {mentions} : {},
      ...upload_id ? {upload_id} : {},
      ...this.props.reply_message_id ? {reply_message_id: this.props.reply_message_id} : {},
      subscription_id: this.props.subscription_id,
    });

    this.setState({ value: '', attachment: null, upload_id: null });
    setTimeout(() => this.calcTextareaHeight());

    if (this.props.reply_message_id) {
      this.props.clearReplyMessage();
    }
  };

  onSendButtonClick = () => {
    if (this.state.isFileLoading) {
      return;
    }

    this.textareaRef.focus();
    const text = this.getFilteredMessage(this.state.value);
    const attachment = this.state.attachment;
    const upload_id = this.state.upload_id;
    const mentions = this.parseMentions(text);

    if (!text && !attachment) {
      this.props.showNotification('No data to send');
      return;
    }

    if (this.props.edit_message_id) {
      this.editMessage({ text, attachment, upload_id });
      return;
    }

    this.sendMessage({
      text,
      mentions,
      upload_id,
      attachment,
    });
  };

  calcTextareaHeight = () => {
    if (!this.textareaRef) {
      return;
    }

    this.textareaRef.style.height = '20px';

    if (this.textareaRef.scrollHeight > 20) {
      this.textareaRef.style.height = this.textareaRef.scrollHeight + 10 + 'px';
      this.inputWrapperRef.style.marginTop =  '10px';
    } else {
      this.inputWrapperRef.style.marginTop =  0;
    }
  };

  isSendButtonShown = () => {
    if (this.state.isFileLoading) {
      return false;
    }

    if (this.state.value) {
      return true;
    }

    if (this.state.attachment) {
      return true;
    }

    return false;
  };

  updateDraft = value => this.props.updateDraft({ id: this.props.subscription_id, value });
  throttleUpdateDraft = throttle(this.updateDraft, 500);

  onInput = event => {
    this.calcTextareaHeight();
    this.setState({ value: event.target.value });

    if (!this.props.edit_message_id && !this.props.reply_message_id) {
      setTimeout(() => this.throttleUpdateDraft(this.state.value));
    }
  };

  closeMessage = () => {
    if (this.props.edit_message_id) {
      this.props.clearEditMessage();
    }

    if (this.props.reply_message_id) {
      this.props.clearReplyMessage();
    }

    this.setState({
      value: this.props.draft ? this.props.draft : '',
      attachment: null,
    });
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.editing_message && !isEqual(this.props.editing_message, nextProps.editing_message)) {
      this.setState({
        ...nextProps.editing_message.text ? {value: nextProps.editing_message.text} : {},
        ...nextProps.editing_message.attachment ? {attachment: nextProps.editing_message.attachment} : {},
      });

      setTimeout(() => this.textareaRef.focus());
    }

    if (this.props.subscription_id !== nextProps.subscription_id) {
      this.setState({
        value: nextProps.draft ? nextProps.draft : '',
        attachment: null,
      });

      setTimeout(() => this.textareaRef.focus());
    }
  }

  componentDidMount() {
    this.calcTextareaHeight();
    this.textareaRef.focus();
    window.addEventListener('keydown', this.handleDocumentKeyDown);
  }

  componwntWillUnmount() {
    document.removeEventListener('keydown', this.handleDocumentKeyDown);
  }

  render() {
    const isSendButtonShown = this.isSendButtonShown();
    const isAttachmentImage = this.state.attachment && this.state.attachment.content_type.match('image/');
    const messageId = this.props.reply_message_id || this.props.edit_message_id;
    const sendButtonName = this.props.edit_message_id ? this.props.t('edit') : this.props.t('send');

    return <div className={cx('input', this.props.className)}>
      <Button
        appearance="_icon-transparent"
        icon="attach"
        onClick={this.attachFile}
        className={style.attach}
      />

      <input
        className={style.attach_input}
        type="file"
        ref={node => this.attachInputRef = node}
        onChange={this.onAttachFileChange}
      />

      <div className={style.section}>
        {messageId &&
          <Message className={style.message} id={messageId} onClose={this.closeMessage} />
        }

        <div className={style.input_wrapper} ref={node => this.inputWrapperRef = node}>
          <div className={style.input_content}>
            <textarea
              placeholder={this.props.t('message')}
              ref={node => this.textareaRef = node}
              value={this.state.value}
              onInput={this.onInput}
              onChange={() => {}}
              onKeyDown={this.onTextareaKeyDown}
            />

            {this.state.attachment &&
              <div
                className={style.preview}
                {...isAttachmentImage ? { style: {'--image': `url(${this.state.attachment.preview})`} } : {}}
              >
                {!isAttachmentImage &&
                  <Icon name="attach" />
                }

                <button onClick={this.resetAttachment}>
                  <Icon name="close" />
                </button>

                <Loading className={style.file_loading} isShown={this.state.isFileLoading} />
              </div>
            }
          </div>

          <button onClick={this.onSendButtonClick} className={cx({'_is-shown': isSendButtonShown})}>
            {sendButtonName}
          </button>
        </div>
      </div>
    </div>;
  }
}

export default compose(
  withNamespaces('translation'),

  connect(
    (state, props) => ({
      draft: get(state.subscriptions.list[props.subscription_id], 'draft', ''),
      edit_message_id: state.messages.edit_message_id,
      reply_message_id: state.messages.reply_message_id,
      users_ids: state.users.ids,
      users_list: state.users.list,
      isMobile: state.device === 'touch',
    }),

    {
      updateDraft: inputActions.updateDraft,
      sendMessage: inputActions.sendMessage,
      updateMessage: inputActions.updateMessage,
      clearEditMessage: messagesActions.clearEditMessage,
      clearReplyMessage: messagesActions.clearReplyMessage,
      showNotification: notificationActions.showNotification,
    },
  ),

  connect(
    (state, props) => ({
      editing_message: props.edit_message_id ? state.messages.list[props.edit_message_id] : null,
    }),
  ),
)(MessageInput);
