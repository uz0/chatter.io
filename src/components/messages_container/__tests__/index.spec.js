import React from 'react';
import { shallow } from 'enzyme';
import { Messages } from '@/components/messages_container';
import data from '@/__tests__/__data__';

describe('Messages component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Messages
      t={() => {}}
      setError={() => {}}
      params={{}}
      currentUser={data.currentUser}
      details={data.details}
      isGalleryOpen={false}
      subscriptions_ids={[]}
    />);

    expect(wrapper).toMatchSnapshot();
  });
});
