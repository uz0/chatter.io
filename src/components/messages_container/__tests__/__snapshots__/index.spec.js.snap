// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Messages
    currentUser={
      Object {
        "auth_token": "Qli089GxuOrWGAAVYzcVJH1px_lRHO6Wf0Ke_icqf8kpndF0Fz_b02li8",
        "avatar": Object {
          "full": "https://chat.mainnetwork.io/rails/active_storage/b…9eb29647350acdf569189259ae78b83/inline_image.jpeg",
          "small": "https://chat.mainnetwork.io/rails/active_storage/r…e8e3e554111dab5bd87656de36a7063/inline_image.jpeg",
        },
        "confirmed_at": "2018-10-16T12:06:32Z",
        "email": "efim1382@gmail.com",
        "id": 55,
        "is_confirmed": true,
        "nick": "efim1382",
        "searchable_nick": true,
      }
    }
    details={
      Object {
        "created_at": "2018-10-17T14:15:05Z",
        "draft": "",
        "group": Object {
          "created_at": "2018-10-17T14:15:05Z",
          "icon": null,
          "id": 5,
          "invite_code": null,
          "is_deleted": false,
          "name": null,
          "participants": Array [
            Object {
              "created_at": "2018-10-17T14:15:05Z",
              "draft": "",
              "group_id": 5,
              "id": 69,
              "last_mentioned_in_message_id": null,
              "last_read_message_id": 5705,
              "mute_until": null,
              "role": "rw",
              "tags": null,
              "user": Object {
                "auth_token": "Qli089GxuOrWGAAVYzcVJH1px_lRHO6Wf0Ke_icqf8kpndF0Fz_b02li8",
                "avatar": Object {
                  "full": "https://chat.mainnetwork.io/rails/active_storage/b…9eb29647350acdf569189259ae78b83/inline_image.jpeg",
                  "small": "https://chat.mainnetwork.io/rails/active_storage/r…e8e3e554111dab5bd87656de36a7063/inline_image.jpeg",
                },
                "confirmed_at": "2018-10-16T12:06:32Z",
                "email": "efim1382@gmail.com",
                "id": 55,
                "is_confirmed": true,
                "nick": "efim1382",
                "searchable_nick": true,
              },
              "user_id": 55,
            },
            Object {
              "created_at": "2018-10-17T14:15:05Z",
              "draft": "",
              "group_id": 5,
              "id": 70,
              "last_mentioned_in_message_id": null,
              "last_read_message_id": 2482,
              "mute_until": null,
              "role": "rw",
              "user": Object {
                "avatar": Object {
                  "full": "https://chat.mainnetwork.io/rails/active_storage/blobs/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBIdz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--33756fcf602882ff40d7f4dc9fac66adbc1345c8/inline_image.png",
                  "small": "https://chat.mainnetwork.io/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBIdz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--33756fcf602882ff40d7f4dc9fac66adbc1345c8/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9MY21WemFYcGxTU0lOTWpBd2VESXdNRDRHT2daRlZBPT0iLCJleHAiOm51bGwsInB1ciI6InZhcmlhdGlvbiJ9fQ==--5bd2204cbe8e3e554111dab5bd87656de36a7063/inline_image.png",
                },
                "id": 59,
                "is_confirmed": false,
                "nick": "ivan ivanov",
              },
              "user_id": 55,
            },
          ],
          "type": "private_chat",
          "updated_at": "2018-10-17T14:15:05Z",
        },
        "group_id": 5,
        "id": 69,
        "last_mentioned_in_message_id": null,
        "last_read_message_id": 5705,
        "mute_until": null,
        "role": "rw",
        "tags": null,
        "user": Object {
          "auth_token": "Qli089GxuOrWGAAVYzcVJH1px_lRHO6Wf0Ke_icqf8kpndF0Fz_b02li8",
          "avatar": Object {
            "full": "https://chat.mainnetwork.io/rails/active_storage/b…9eb29647350acdf569189259ae78b83/inline_image.jpeg",
            "small": "https://chat.mainnetwork.io/rails/active_storage/r…e8e3e554111dab5bd87656de36a7063/inline_image.jpeg",
          },
          "confirmed_at": "2018-10-16T12:06:32Z",
          "email": "efim1382@gmail.com",
          "id": 55,
          "is_confirmed": true,
          "nick": "efim1382",
          "searchable_nick": true,
        },
        "user_id": 55,
      }
    }
    isGalleryOpen={false}
    params={Object {}}
    setError={[Function]}
    subscriptions_ids={Array []}
    t={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <withRouter(Connect(Header))
          chatId={69}
          className="header"
        />,
        <div
          className="list"
          id="messages-scroll"
        >
          <p
            className="empty"
          />
          <Loading
            className="loading"
            isShown={true}
          />
        </div>,
        <Connect(LoadNamespace(Connect(Wrapped)))
          className="typings"
          subscription_id={69}
        />,
        <LoadNamespace(Connect(Connect(MessageInput)))
          className="input"
          scrollListMessagesToBottom={[Function]}
          subscription_id={69}
        />,
      ],
      "className": "messages",
      "data-e2c53d2c": true,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "chatId": 69,
          "className": "header",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            false,
            <p
              className="empty"
            />,
            <Loading
              className="loading"
              isShown={true}
            />,
          ],
          "className": "list",
          "id": "messages-scroll",
        },
        "ref": [Function],
        "rendered": Array [
          undefined,
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "empty",
            },
            "ref": null,
            "rendered": null,
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "loading",
              "isShown": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "typings",
          "subscription_id": 69,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "input",
          "scrollListMessagesToBottom": [Function],
          "subscription_id": 69,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <withRouter(Connect(Header))
            chatId={69}
            className="header"
          />,
          <div
            className="list"
            id="messages-scroll"
          >
            <p
              className="empty"
            />
            <Loading
              className="loading"
              isShown={true}
            />
          </div>,
          <Connect(LoadNamespace(Connect(Wrapped)))
            className="typings"
            subscription_id={69}
          />,
          <LoadNamespace(Connect(Connect(MessageInput)))
            className="input"
            scrollListMessagesToBottom={[Function]}
            subscription_id={69}
          />,
        ],
        "className": "messages",
        "data-e2c53d2c": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "chatId": 69,
            "className": "header",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              false,
              <p
                className="empty"
              />,
              <Loading
                className="loading"
                isShown={true}
              />,
            ],
            "className": "list",
            "id": "messages-scroll",
          },
          "ref": [Function],
          "rendered": Array [
            undefined,
            false,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "className": "empty",
              },
              "ref": null,
              "rendered": null,
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "loading",
                "isShown": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "typings",
            "subscription_id": 69,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "input",
            "scrollListMessagesToBottom": [Function],
            "subscription_id": 69,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
